src/term.rs

```diff
@@ -98,18 +106,46 @@ impl<VS: VersionSet> Term<VS> {
         }
     }
 
+    /// Check whether two terms are mutually exclusive.
+    ///
+    /// An optimization for the native implementation of checking whether the intersection of two sets is empty.
+    pub(crate) fn is_disjoint(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Positive(r1), Self::Positive(r2)) => r1.is_disjoint(r2),
+            (Self::Negative(r1), Self::Negative(r2)) => r1 == &VS::empty() && r2 == &VS::empty(),
+            // If the positive term is a subset of the negative term, it lies fully in the region that the negative
+            // term excludes.
+            (Self::Positive(r1), Self::Negative(r2)) => r1.subset_of(r2),
+            // Inversely, if there is a region outside the negative, they overlap in this region.
+            (Self::Negative(r1), Self::Positive(r2)) => r2.subset_of(r1),
+        }
+    }
+
     /// Compute the union of two terms.
     /// If at least one term is negative, the union is also negative.
     pub(crate) fn union(&self, other: &Self) -> Self {
-        (self.negate().intersection(&other.negate())).negate()
+        match (self, other) {
+            (Self::Positive(r1), Self::Positive(r2)) => Self::Positive(r1.union(r2)),
+            (Self::Positive(r1), Self::Negative(r2)) => {
+                Self::Negative(r1.union(&r2.complement()).complement())
+            }
+            (Self::Negative(r1), Self::Positive(r2)) => {
+                Self::Negative(r1.complement().union(r2).complement())
+            }
+            (Self::Negative(r1), Self::Negative(r2)) => Self::Negative(r1.intersection(r2)),
+        }
     }
 
     /// Indicate if this term is a subset of another term.
     /// Just like for sets, we say that t1 is a subset of t2
     /// if and only if t1 ∩ t2 = t1.
-    #[cfg(test)]
     pub(crate) fn subset_of(&self, other: &Self) -> bool {
-        self == &self.intersection(other)
+        match (self, other) {
+            (Self::Positive(r1), Self::Positive(r2)) => r1.subset_of(r2),
+            (Self::Positive(r1), Self::Negative(r2)) => r1.subset_of(&r2.complement()),
```

Comment:
`r1.subset_of(&r2.complement())` is `r1.is_disjoint(r2)`. This saves `complement` with its associated allocations and drop. On the Sudoku case I see an additional 30% drop to 20s!